File: CPP/ByTier/B/1687_B_Railway_System.cpp
// Problem: CF 1687 B - Railway System
// https://codeforces.com/contest/1687/problem/B

/*
Algorithm: Binary search + Greedy + Sorting
Approach:
1. First, we determine the length of each edge by querying with all edges set to 1,
   except one edge which is set to 0. This allows us to find the value of every edge.
2. Then, we sort edges based on their lengths in ascending order.
3. We simulate adding edges one by one in this sorted order.
   For each edge, we test if including it increases the maximum capacity.
   If it does, we include it and update the answer.
   Otherwise, we exclude it.

Time Complexity: O(m^2 * log(m)) due to sorting and nested queries.
Space Complexity: O(m) for storing edges and queries.

This solution uses a greedy approach combined with binary search over the edge values
to efficiently determine the minimum capacity of the railway system, which is the
maximum value of a spanning forest under the assumption of optimal edge inclusion.
*/

#include <algorithm>
#include <iostream>

using namespace std;

int n, m, e[510], p[510], ans;
char s[500010];

// Comparator to sort edges by their lengths
bool cmp(const int x, const int y) { return e[x] < e[y]; }

int main() {
  cin >> n >> m;
  // Initialize string s with '0' for all edges
  for (int i = 1; i <= m; i++)
    s[i] = '0', p[i] = i;

  // Query to get the value of each edge
  // For each edge, set it to 1 and others to 0 to isolate its contribution
  for (int i = 1; i <= m; i++) {
    s[i] = '1';                    // Activate current edge
    cout << "? " << s + 1 << endl; // Send query to interactor
    s[i] = '0';                    // Deactivate current edge
    cin >> e[i];                   // Receive result from interactor
  }

  // Sort edge indices by their lengths (ascending order)
  sort(p + 1, p + m + 1, cmp);

  // Greedily add edges in ascending order to build the spanning forest
  for (int i = 1, opt; i <= m; ++i) {
    s[p[i]] = '1';                 // Activate the i-th smallest edge
    cout << "? " << s + 1 << endl; // Query with current edge set
    cin >> opt;                    // Get maximum capacity with this edge included

    // If the new capacity is greater than previous + edge value,
    // then this edge contributes to the maximum capacity and we accept it
    if (opt - ans != e[p[i]])      // Check if edge was contributing
      s[p[i]] = '0';               // If not, remove it
    else
      ans = opt;                   // Else update the answer
  }

  // Output final result
  cout << "! " << ans << endl;
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1687/1687_B_Railway_System.py
# Problem: CF 1687 B - Railway System
# https://codeforces.com/contest/1687/problem/B

"""
Algorithm: 
This code solves an interactive problem to determine the minimum spanning tree (MST) capacity of a railway system.
The approach uses queries to determine which edges are part of the MST by checking if adding an edge increases the maximum capacity.
It first queries each edge individually to obtain the maximum capacity when only that edge is active. Then it uses these values
to find which edges should be included in the MST.

Time Complexity: O(m^2)
Space Complexity: O(m)

The algorithm works by:
1. First, querying each edge independently to get the impact of that edge on the maximum capacity.
2. Then, iteratively determining whether to include each edge in the final MST by checking if its inclusion affects
   the overall capacity.
3. Finally, outputting the minimum capacity of the full spanning forest, which is the total weight of the MST.

Note: This solution uses at most 2*m queries as required.
"""

n, m = map(int, input().split())
a = []
for i in range(m):
    # Query each edge individually to determine its impact on maximum capacity
    print("?", "0" * i + "1" + "0" * (m - i - 1), flush=1)
    a.append(int(input()))

cur = 0
s = ["0" for i in range(m)]  # s represents the current set of active edges

# Try to build the MST by adding edges one by one
for i in range(m):
    # Find the edge with the smallest impact on maximum capacity (i.e., least important edge)
    x = 0
    for j in range(m):
        if a[x] > a[j]:
            x = j
    
    # Set the selected edge as active in the current solution
    s[x] = "1"
    print("? ", *s, sep="", flush=1)
    c = int(input())
    
    # If including this edge actually increases the maximum capacity, we include it in the MST
    if cur + a[x] == c:
        cur += a[x]
    else:
        # Otherwise, this edge is not needed in the MST, so we unset it
        s[x] = "0"
    
    # Mark the edge as processed by setting its value to a large number
    a[x] = 2000000

# Output the minimum capacity (which is the MST weight)
print("!", cur, flush=1)


# https://github.com/VaHiX/CodeForces/